docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1058 @@
+        }
+    }
+
+    // If any elements remain, process the last vector in the search space.
+    if (buffer.Length % Vector128<int>.Count != 0)
+    {
+        loaded = Vector128.LoadUnsafe(ref searchSpace, oneVectorAwayFromEnd);
+        if (Vector128.Equals(loaded, values) != Vector128<int>.Zero)
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+```
+
+`Vector128.Create(value)` creates a new vector with all elements initialized to the specified value. So `Vector128<int>.Zero` is an equivalent of `Vector128.Create(0)`.
+
+`Vector128.Equals(Vector128 left, Vector128 right)` compares two vectors and returns a vector whose elements are all-bits-set or zero, depending on if the provided elements in left and right were equal. If the result of comparison is non zero, it means that there was at least one match.
+
+### Access violation (AV) testing
+
+Handling the remainder in an invalid way, may lead to non-deterministic and hard to diagnose issues.
+
+Let's look at the following code:
+
+```diff
+nuint elementOffset = 0;
+while (elementOffset < (nuint)buffer.Length)
+{
+    loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset);
+
+    elementOffset += (nuint)Vector128<int>.Count;
+}
+```
+
+How many time the loop is going to execute for a buffer of six integers? Twice! The first time it's going to load the first four elements, the second time it's going to load the two last elements and turn random memory that is following the buffer into next two elements!
+
+Writing tests that detect such issues is hard, but not impossible. .NET Team uses a helper utility called [BoundedMemory](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Creation.cs) that allocates memory region which is immediately preceded by or immediately followed by a poison (`MEM_NOACCESS`) page. Attempting to read the memory immediately before or after it results in `AccessViolationException`.
+
+## Loading and storing vectors
+
+### Loading
+
+Both `Vector128` and `Vector256` provide at least five ways of loading them from memory:
+
+```cs
+public static class Vector128
+{
+    public static Vector128<T> Load<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAligned<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAlignedNonTemporal<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadUnsafe<T>(ref T source) where T : struct;
+    public static Vector128<T> LoadUnsafe<T>(ref T source, nuint elementOffset) where T : struct;
+}
+```
+
+The first three overloads require a pointer to the source. To be able to use a pointer in a safe way, the buffer needs to be pinned first (the GC is not tracking unmanaged pointers, we have to ensure that the memory does not get moved by GC in the meantime, as the pointers would silently become invalid). That is simple, the problem is doing the pointer arithmetic right:
```

Comment:
```suggestion
The first three overloads require a pointer to the source. To be able to use a pointer in a safe way, the buffer needs to be pinned first. This is because the GC cannot track unmanaged pointers. It needs help to ensure that it doesn't move the memory while you're using it, as the pointers would silently become invalid. The tricky part here is doing the pointer arithmetic right:
```